server/src/main/java/org/elasticsearch/cluster/metadata/DataStream.java

```diff
@@ -201,4 +204,77 @@ public boolean equals(Object o) {
     public int hashCode() {
         return Objects.hash(name, timeStampField, indices, generation);
     }
+
+    public static final class TimestampField implements Writeable, ToXContentObject {
```

Comment:
Do we want to allow this field to contain any mapping properties or just the field's data type? Also, should we require that it contains a data type property of either `date` or `date_nanos`?

Reply:
This class should contain the mapping attributes. Part of the mapping attributes is the `type` attribute, so that data type is kept track of.

Reply:
Ok, should we permit _not_ specifying a data type? A data type of either `date` or `date_nanos` is required on the composable template, so it seems like maybe it should be required on the data stream, too?

Reply:
It isn't possible that any other data type than `date` and `date_nanos` is being included the `fieldMapping` map, because no composable index template with a data stream definition can be created with timestamp field mapping that is not of type `date` or `date_nanos`.

When creating a data stream the contents of the field mapping is copied and this includes a `type` attribute, so that we know whether either `date` or `date_nanos` is specified.

I can add an assert here that checks that value for the `type` key either returns `date` or `date_nanos` for sanity checking.